### PR TITLE
netspeed: Show all IPv4 and IPv6 addresses in tooltip

### LINKED ISSUE #557: netspeed: support IPv6 in tooltip window
#### Expected behaviour

When hovering the applet in the status bar, it should show the IPv6 address of the NIC in the tooltip window.

#### Actual behaviour

When hovering the applet in the status bar, it only shows the legacy IPv4 address. If no legacy address is configured on that interface, it will only show "0.0.0.0" in the tooltip window.

#### Steps to reproduce the behaviour

Configure interface with IPv6 address. Add netspeed applet to taskbar. Hover it with the mouse for the tooltip window to appear.

#### MATE general version

MATE Desktop Environment 1.24.0

#### Package version

MATE Netspeed 1.24.0

#### Linux Distribution

Mint 20 MATE 64-bit

### PR DESCRIPTION
**Pull request by treysis**
Fixes #557.

### CODE REVIEW

**Comment by treysis** [comment]
Argh, how do I get rid of all the previous commits? Always when I try to rebase it will delete all my previous changes....

**Comment by vkareh** [comment]
Do you want only the last commit? Try doing this: `git rebase -i HEAD~4` and follow the instructions to `drop` commits from that list. Save and exit, and you will then have only the remaining commit

**Comment by treysis** [comment]
I don't know. I thought I rebased to the current master branch, created a new branch for my fix, did my adjustments to the code, pushed it, opened pull request. No idea why it shows the other commits in this PR. I'll better leave it as it is now.

**Comment by vkareh** [comment]
let's fix it! what do you see when you do `git remote -v`?

**Comment by treysis** [comment]
```
origin	https://github.com/treysis/mate-applets (fetch)
origin	https://github.com/treysis/mate-applets (push)
```

**Comment by vkareh** [comment]
huh, how are you updating your master then?
you should add the upstream as a remote: `git remote add upstream git@github.com:mate-desktop/mate-applets.git`
then you can actually update:
```bash
$ git checkout master # checks out your local master
$ git fetch upstream # fetches all latest changes from the upstream remote
$ git pull upstream master # merges all changes from upstream into your local master
$ git push # pushes those changes to your origin branch on github
```
_Now_ you can do `git rebase master` on your `ipv6-tooltip-all` branch. Once you rebase `master`, you'll need to `git push --force` your `ipv6-tooltip-all` so that it overrides all the previous junk

**Comment by treysis** [comment]
> huh, how are you updating your master then?

I did that. But then I got confused about everything and decided to remove my local copy and clone it again. After cloning I just had forgotten to add the upstream repo again.

I'll try your suggestions. Thx!

File: netspeed/src/netspeed.c
Comment (rbuj) [inline netspeed/src/netspeed.c:1539]: You can remove `ipv6` from `devinfo`.
```diff
diff --git a/netspeed/src/backend.c b/netspeed/src/backend.c
index 7701b7c0..dc7f130d 100644
--- a/netspeed/src/backend.c
+++ b/netspeed/src/backend.c
@@ -400,7 +400,6 @@ get_device_info (const char  *device,
     if (devinfo->running) {
         devinfo->ip = netload.address;
         devinfo->netmask = netload.subnet;
-        memcpy (devinfo->ipv6, netload.address6, 16);
 #if defined (HAVE_NL)
         if (devinfo->type != DEV_WIRELESS) {
             devinfo->tx = netload.bytes_out;
diff --git a/netspeed/src/backend.h b/netspeed/src/backend.h
index c67c9b39..651b9898 100644
--- a/netspeed/src/backend.h
+++ b/netspeed/src/backend.h
@@ -61,7 +61,6 @@ typedef struct {
     guint32        netmask;
     guint32        ptpip;
     guint8         hwaddr [8]; /* EUI-48 or EUI-64 */
-    guint8         ipv6 [16];
     char          *essid;
     gboolean       up;
     gboolean       running;
```

**Comment by rbuj** [comment]
tip: before starting a new pr, first update your local and remote copy as @vkareh pointed.
```
git clone git@github.com:treysis/mate-applets.git
cd mate-applets
git remote add upstream https://github.com/mate-desktop/mate-applets.git
git fetch upstream
git checkout master
git merge upstream/master
git push
```
note: you only have to do the steps 1 and 3 once

**Comment by lukefromdc** [comment]
Since I have no landline and only connect by tethering I don't have a way to test an IPv6 address.

**Comment by treysis** [comment]
IPv6 works also with tethering, if your carrier supports it.

Anyways, for testing you can just rely on link-local addresses or you can just configure dummy addresses as static addresses and check it out.

**Comment by lukefromdc** [comment]
I can never call customer service as I am keeping my name from them(prepaid plans only). I don't know enough about networking to really experiment with this, didn't even have the Internet at home when I started playing with software.

**Comment by treysis** [comment]
I cleaned it up and rebased to latest `master` (as of now).

File: netspeed/src/netspeed.c
Comment (rbuj) [inline netspeed/src/netspeed.c:1549]: increase the indent

File: netspeed/src/netspeed.c
Comment (rbuj) [inline netspeed/src/netspeed.c:1553]: same here

**Review by rbuj — COMMENTED** [review]
fix indent. you should also add a white space between the function name and the parentheses

File: netspeed/data/org.mate.panel.applet.netspeed.gschema.xml.in
Comment (rbuj) [inline netspeed/data/org.mate.panel.applet.netspeed.gschema.xml.in:19]: i don't think it should be enabled by default

File: netspeed/src/backend.c
Comment (rbuj) [inline netspeed/src/backend.c:254]: It will display mixed ipv4 and ipv6 addresses

File: netspeed/data/org.mate.panel.applet.netspeed.gschema.xml.in
Comment (treysis) [reply netspeed/data/org.mate.panel.applet.netspeed.gschema.xml.in:19]: I still think it should. IPv6 is the new normal. Whoever disables it or doesn't want to use it has to put extra effort.

File: netspeed/src/backend.c
Comment (treysis) [reply netspeed/src/backend.c:254]: Hm, you might be right. In my cases it always showed IPv4 before IPv6, I guess because of the colon-notation instead of point-notation. What's your suggestion here? Remove it?

Comment (rbuj) [reply netspeed/src/backend.c:254]: > Hm, you might be right. In my cases it always showed IPv4 before IPv6, I guess because of the colon-notation instead of point-notation. What's your suggestion here? Remove it?

No, you can use another GCompareFunc.

Comment (rbuj) [reply netspeed/src/backend.c:254]: you can also use two lists...

Comment (treysis) [reply netspeed/src/backend.c:254]: You are right. I checked it and the reason IPv4 mostly appear in front is because most IPv6 addresses start with >=2, while most IPv4 addresses start with <=2. Now I am thinking of 3 solutions:
1. Use different sort function (do you have a specific one in mind?)
2. Use 2 separate lists which are filled by 2 separate functions (`get_ipv4_address_list` and `get_ipv6_address_list`)
3. Modify `get_ip_address_list` to return structs that contain `ip`, `netmask`, and `scope` and then use this data in `netspeed.c`.

However, I would postpone option 3 because it is a much bigger code change and I don't know when I will find the time for this.

Comment (rbuj) [reply netspeed/src/backend.c:254]: To make your choice easier, consider that at most there is an ipv4 address and it's stored in binary format in device->ip variable (2nd solution)

File: netspeed/src/backend.c
Comment (rbuj) [inline netspeed/src/backend.c:227]: you can combine these 2 clauses in one if statement, i.e.:
`if (!g_strcmp0 (iface->ifa_name, iface_name)) && (iface->ifa_addr->sa_family == AF_INET6))`

Comment (rbuj) [reply netspeed/src/backend.c:227]: or place family check before comparing the interface names like NULL check.

File: netspeed/src/backend.c
Comment (rbuj) [inline netspeed/src/backend.c:236]: Use `g_strdup (ip6)` instead of `g_strdup_printf ("%s", ip6)`

File: netspeed/src/netspeed.c
Comment (rbuj) [inline netspeed/src/netspeed.c:1567]: you can use `g_string_append` instead of `g_string_append_printf`

File: netspeed/src/netspeed.c
Comment (rbuj) [inline netspeed/src/netspeed.c:1556]: you can use `g_string_append` instead of `g_string_append_printf`

**Comment by treysis** [comment]
Still any objections?

**Comment by rbuj** [comment]
I don't think the best option is to set the show-all-addresses key to true by default, since as I mentioned above the ipv6 addressing is not the most used at the moment, moreover it doesn't provide any useful information when an interface is just ipv6 link-local, or loopback. For this reason I cannot approve this pull request. If we omit this objection the other changes seem correct.

**Comment by raveit65** [comment]
I agree to don't show them as default in tooltip, because it's not a big thing to enable it in preferences.
Also , an admin or distro maintainer can always do a gsettings override, which is easy to do, when he think this would be a better default for his distro or let say 1000 machines he have to administrate.
For fedora i have done this at several places ;)
So, please change that to `false` and we can merge it if there are no other concerns.

**Comment by treysis** [comment]
Ok, I changed it.

**Review by rbuj — COMMENTED** [review]
The tooltip displayed that an interface "has no ip", when it uses only ipv6 addressing and the "show all ip addresses on tooltip" option is unchecked.

**Comment by treysis** [comment]
Yes. Should it better read `no IPv4`? I left it this way as it is more in line with how someone unaware of IPv6 would expect and how it used to be in the past.

**Comment by rbuj** [comment]
In the example below the interface has no ipv4 address but has two ipv6 addresses, however in this case it only considers ipv4 addresses.

Untitled

**Comment by treysis** [comment]
Do you suggest that in this case it should show the IPv6 addresses, even if the checkbox is unchecked?

**Comment by rbuj** [comment]
The `get_ip6_address_list` function returns the ipv6 addresses sorted alphabetically, in this case you can display only one ipv6 address, perhaps the most important (https://en.wikipedia.org/wiki/IPv6_address#Special_addresses)

**Comment by treysis** [comment]
I guess it should be the GUA address with Privacy Extensions > Manual > Stable Privacy > EUI-64 (depending on what's available). But this is also somewhat user-dependent. And I am not so convinced that this is a meaningful approach in general. I would rather change the checkbox to read `Show IPv6 addresses on tooltip`, and then change the message from `has no IP` to `no IPv4`.

**Comment by rbuj** [comment]
At least all ipv6 addresses should be displayed if ipv4 address is not used, instead of showing `has no ip`.

**Comment by rbuj** [comment]
`::/128` is equivalent to `has no ip` in ipv6, `strlen (ipv6_text)` in the old code.

**Comment by treysis** [comment]
> At least all ipv6 addresses should be displayed if ipv4 address is not used, instead of showing `has no ip`.

I went with this approach for now as it was quite easy to implement. IMO this should be merged to provide at least a very basic IPv6 support. I am open to refining this in the future!

File: netspeed/src/netspeed.c
Comment (rbuj) [inline netspeed/src/netspeed.c:1406]: New runtime warning:
```
(mate-netspeed-applet:299757): GLib-CRITICAL **: 02:10:13.790: g_string_free: assertion 'string != NULL' failed
```

File: netspeed/src/netspeed.c
Comment (rbuj) [inline netspeed/src/netspeed.c:1403]: you use the value of a variable (`!string`) after freeing it (`g_string_free (string, TRUE);`)

Comment (treysis) [reply netspeed/src/netspeed.c:1403]: It's not always freed. And it works that way without errors and the tooltip is displayed correctly.

Comment (rbuj) [reply netspeed/src/netspeed.c:1403]: Test: only ipv4 addressing

It floods the log when the cursor is over the applet:

test-treysis

Comment (rbuj) [reply netspeed/src/netspeed.c:1403]:
```
$ LANG=C journalctl --no-pager --boot | grep mate-netspeed
<cut>
Jan 08 12:31:49 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:50 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:51 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:52 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:53 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:54 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:55 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:56 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:57 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:58 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:31:59 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:32:00 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:32:01 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
Jan 08 12:32:02 localhost.localdomain mate-netspeed-a[121170]: g_string_free: assertion 'string != NULL' failed
```

Comment (treysis) [reply netspeed/src/netspeed.c:1403]: Weird. It didn't for me when I tested it yesterday. I'll habe a look at it again.

Comment (treysis) [reply netspeed/src/netspeed.c:1403]: @rbuj Are you sure you checked out the right commit? I can't reproduce this.

Comment (rbuj) [reply netspeed/src/netspeed.c]: Your last update seems to fix the issue:
```diff
diff --git a/netspeed/src/netspeed.c b/netspeed/src/netspeed.c
index 91555d4d..2dd1e601 100644
--- a/netspeed/src/netspeed.c
+++ b/netspeed/src/netspeed.c
@@ -1388,6 +1388,10 @@ update_tooltip (NetspeedApplet *netspeed)
             g_slist_free_full (ip6_address_list, g_free);
         }
 
+        if (!netspeed->devinfo->ip && !string) {
+            g_string_append (tooltip, "has no ip");
+        }
+
         if (netspeed->devinfo->ip) {
             format_ipv4 (netspeed->devinfo->ip, ipv4_text);
             g_string_append (tooltip, ipv4_text);
@@ -1397,14 +1401,9 @@ update_tooltip (NetspeedApplet *netspeed)
             g_string_append_printf (tooltip,
                                     _("\n%s"),
                                     string->str);
+            g_string_free (string, TRUE);
         }
 
-        if (!netspeed->devinfo->ip && !string) {
-            g_string_append (tooltip, "has no ip");
-        }
-
-        g_string_free (string, TRUE);
-
         if (netspeed->show_sum) {
             g_string_append_printf (tooltip,
                                     _("\nin: %s out: %s"),
```

**Review by rbuj — APPROVED** [review]
It works fine, although it always displays ipv6 addresses from the second line of the tooltip.

**Comment by treysis** [comment]
> it always displays ipv6 addresses from the second line of the tooltip

That is intentional. Since IPv6 addresses are usually longer, it would make the tooltip window even wider. And imo it looks nicer if all IPv6 addresses have the same indentation, i.e. beginning of the line.

**Comment by treysis** [comment]
Can we get this merged?

**Comment by treysis** [comment]
Rebased to current master....thx for teaching me git 😊

**Comment by raveit65** [comment]
So many changes with netspeed-applet in the last weeks, but showing the important ip address is missing. ;)
The result of `curl ifconfig.me`
Very important when you use a vpn connection.

**Review by raveit65 — APPROVED** [review]
Builds and runs fine.

**Pull request merged**
